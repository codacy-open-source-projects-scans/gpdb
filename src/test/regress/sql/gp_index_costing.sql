--
-- Test Cases to check Index scan, Index Only scan,
-- Dynamic Index scan & Dynamic Index only scan costing
--


---------------------------------------------------------------------
-- Scenario1: Verify impact of unused index column in predicate,
-- which helps in selecting best index in a multiple index scenario.
---------------------------------------------------------------------

    -- Case A : Index - idx_ab, Columns used in predicate : a
    -- Case B : Index - idx_ab, Columns used in predicate : b
    -- In the above two cases, the index scan using idx_ab should be higher
    -- in CASE B, compared for CASE A, as the in Case-B, index column
    -- of higher significance is unused  i.e. col 'a'

drop table if exists foo;
create table foo (a int , b int, c int) distributed by (a);
insert into foo select i,i,i from generate_series(1,5)i;
-- Adding 1,1,1 in the table so that we have different values for NDV and table rows.
insert into foo select 1,1,1*i/i from generate_series(1,5)i;

-- 1.1 Test case for index scans
CREATE INDEX idx_foo_ab ON foo USING btree(a,b);
CREATE INDEX idx_foo_ba ON foo USING btree(b,a);
analyze foo;
    -- index idx_foo_ab should be selected
explain select * from foo where a=1;
    -- Index idx_foo_ba should be selected
explain select * from foo where b=1;
drop index idx_foo_ab;
drop index idx_foo_ba;

-- 1.2 Test case for index only scans
CREATE INDEX idx_foo_abc ON foo USING btree(a,b,c);
CREATE INDEX idx_foo_cba ON foo USING btree(c,b,a);
vacuum analyze foo;
    --  Index idx_foo_abc should be selected
explain select * from foo where a=1;
    --  Index idx_foo_cba should be selected
explain select * from foo where c=1;
drop index idx_foo_abc;
drop index idx_foo_cba;


-- 1.3 Test case for dynamic index scans
drop table if exists foo;
drop table if exists bar_PT;
create table foo (a int , b int, c int) distributed by (a);
insert into foo select i,i,i from generate_series(1,5)i;
analyze foo;
    -- Partitioned Table
create table bar_PT (a int, b int, c int) partition by range(a) (start (1) inclusive end (12) every (2)) distributed by (a);
insert into bar_PT select i,i,i from generate_series(1,11)i;
CREATE INDEX idx_bar_PT_ab ON bar_PT USING btree(a,b);
CREATE INDEX idx_bar_PT_ba ON bar_PT USING btree(b,a);
analyze bar_PT;
    --  Index idx_bar_PT_ab should be selected
explain select * from bar_PT join foo on bar_PT.a =foo.a;
    --  Index idx_bar_PT_ba should be selected
explain select * from bar_PT join foo on bar_PT.b =foo.b;
drop index idx_bar_PT_ab;
drop index idx_bar_PT_ba;

-- 1.4 Test case for dynamic index only scans
CREATE INDEX idx_bar_PT_abc ON bar_PT USING btree(a,b,c);
CREATE INDEX idx_bar_PT_cba ON bar_PT USING btree(c,b,a);
vacuum analyze bar_PT;
    --  Index idx_bar_PT_abc should be selected
explain select * from bar_PT join foo on bar_PT.a =foo.a;
    --  Index idx_bar_PT_cba should be selected
explain select * from bar_PT join foo on bar_PT.c =foo.c;
drop table if exists foo;
drop table if exists bar_PT;

------------------------------------------------
-- Scenario2: Unindexed predicate column in index.
------------------------------------------------
-- If the index does not cover all the columns of the predicate, than a scan cost
-- using it should be higher compared to an index which covers all the predicate columns.
drop table if exists foo;
create table foo (a int , b int, c int) distributed by (a);
insert into foo select i,i,i from generate_series(1,5)i;
-- Adding 1,1,1 in the table so that we have different values for NDV and table rows.
insert into foo select 1,1,1*i/i from generate_series(1,5)i;

-- 2.1 Test case for index scans
CREATE INDEX idx_foo_a ON foo USING btree(a);
CREATE INDEX idx_foo_abc ON foo USING btree(a,b,c);
vacuum analyze foo;

-- Query1 - Index idx_foo_a should be selected.
explain select * from foo where a=1;
-- Query2 - Index idx_foo_abc should be selected.
explain select * from foo where a=1 and b=1 and c=1;
drop index idx_foo_a;
drop index idx_foo_abc;

-- 2.2 Test case for dynamic index scans
drop table if exists foo;
drop table if exists bar_PT;
create table foo (a int , b int, c int) distributed by (a);
insert into foo select i,i,i from generate_series(1,5)i;
analyze foo;
    -- Partitioned Table
create table bar_PT (a int, b int, c int) partition by range(a) (start (1) inclusive end (12) every (2)) distributed by (a);
insert into bar_PT select i,i,i from generate_series(1,11)i;
CREATE INDEX idx_bar_PT_a ON bar_PT USING btree(a);
CREATE INDEX idx_bar_PT_abc ON bar_PT USING btree(a,b,c);
vacuum analyze bar_PT;

-- Query1: Index idx_bar_PT_a should be selected.
explain select * from bar_PT join foo on bar_PT.a =foo.a ;
-- Query2 : Index idx_bar_PT_abc should be selected.
explain select * from bar_PT join foo on bar_PT.a =foo.a and bar_PT.b =foo.b and bar_PT.c =foo.c;
drop table if exists foo;
drop table if exists bar_PT;
