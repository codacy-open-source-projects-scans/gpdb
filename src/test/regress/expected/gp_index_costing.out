--
-- Test Cases to check Index scan, Index Only scan,
-- Dynamic Index scan & Dynamic Index only scan costing
--
---------------------------------------------------------------------
-- Scenario1: Verify impact of unused index column in predicate,
-- which helps in selecting best index in a multiple index scenario.
---------------------------------------------------------------------
    -- Case A : Index - idx_ab, Columns used in predicate : a
    -- Case B : Index - idx_ab, Columns used in predicate : b
    -- In the above two cases, the index scan using idx_ab should be higher
    -- in CASE B, compared for CASE A, as the in Case-B, index column
    -- of higher significance is unused  i.e. col 'a'
drop table if exists foo;
NOTICE:  table "foo" does not exist, skipping
create table foo (a int , b int, c int) distributed by (a);
insert into foo select i,i,i from generate_series(1,5)i;
-- Adding 1,1,1 in the table so that we have different values for NDV and table rows.
insert into foo select 1,1,1*i/i from generate_series(1,5)i;
-- 1.1 Test case for index scans
CREATE INDEX idx_foo_ab ON foo USING btree(a,b);
CREATE INDEX idx_foo_ba ON foo USING btree(b,a);
analyze foo;
    -- index idx_foo_ab should be selected
explain select * from foo where a=1;
                                 QUERY PLAN                                  
-----------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..1.12 rows=6 width=12)
   ->  Seq Scan on foo  (cost=0.00..1.04 rows=2 width=12)
         Filter: (a = 1)
 Optimizer: Postgres-based planner
(4 rows)

    -- Index idx_foo_ba should be selected
explain select * from foo where b=1;
                                 QUERY PLAN                                  
-----------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1.12 rows=6 width=12)
   ->  Seq Scan on foo  (cost=0.00..1.04 rows=2 width=12)
         Filter: (b = 1)
 Optimizer: Postgres-based planner
(4 rows)

drop index idx_foo_ab;
drop index idx_foo_ba;
-- 1.2 Test case for index only scans
CREATE INDEX idx_foo_abc ON foo USING btree(a,b,c);
CREATE INDEX idx_foo_cba ON foo USING btree(c,b,a);
vacuum analyze foo;
    --  Index idx_foo_abc should be selected
explain select * from foo where a=1;
                                 QUERY PLAN                                  
-----------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..1.12 rows=6 width=12)
   ->  Seq Scan on foo  (cost=0.00..1.04 rows=2 width=12)
         Filter: (a = 1)
 Optimizer: Postgres-based planner
(4 rows)

    --  Index idx_foo_cba should be selected
explain select * from foo where c=1;
                                 QUERY PLAN                                  
-----------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1.12 rows=6 width=12)
   ->  Seq Scan on foo  (cost=0.00..1.04 rows=2 width=12)
         Filter: (c = 1)
 Optimizer: Postgres-based planner
(4 rows)

drop index idx_foo_abc;
drop index idx_foo_cba;
-- 1.3 Test case for dynamic index scans
drop table if exists foo;
drop table if exists bar_PT;
NOTICE:  table "bar_pt" does not exist, skipping
create table foo (a int , b int, c int) distributed by (a);
insert into foo select i,i,i from generate_series(1,5)i;
analyze foo;
    -- Partitioned Table
create table bar_PT (a int, b int, c int) partition by range(a) (start (1) inclusive end (12) every (2)) distributed by (a);
insert into bar_PT select i,i,i from generate_series(1,11)i;
CREATE INDEX idx_bar_PT_ab ON bar_PT USING btree(a,b);
CREATE INDEX idx_bar_PT_ba ON bar_PT USING btree(b,a);
analyze bar_PT;
    --  Index idx_bar_PT_ab should be selected
explain select * from bar_PT join foo on bar_PT.a =foo.a;
                                        QUERY PLAN                                         
-------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=1.04..7.23 rows=5 width=24)
   ->  Hash Join  (cost=1.04..7.16 rows=2 width=24)
         Hash Cond: (bar_pt_1_prt_1.a = foo.a)
         ->  Append  (cost=0.00..6.09 rows=6 width=12)
               Partition Selectors: $0
               ->  Seq Scan on bar_pt_1_prt_1  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_2  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_3  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_4  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_5  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_6  (cost=0.00..1.01 rows=1 width=12)
         ->  Hash  (cost=1.02..1.02 rows=2 width=12)
               ->  Partition Selector (selector id: $0)  (cost=0.00..1.02 rows=2 width=12)
                     ->  Seq Scan on foo  (cost=0.00..1.02 rows=2 width=12)
 Optimizer: Postgres-based planner
(15 rows)

    --  Index idx_bar_PT_ba should be selected
explain select * from bar_PT join foo on bar_PT.b =foo.b;
                                            QUERY PLAN                                            
--------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=1.15..7.34 rows=5 width=24)
   ->  Hash Join  (cost=1.15..7.27 rows=2 width=24)
         Hash Cond: (bar_pt_1_prt_1.b = foo.b)
         ->  Append  (cost=0.00..6.09 rows=6 width=12)
               ->  Seq Scan on bar_pt_1_prt_1  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_2  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_3  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_4  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_5  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_6  (cost=0.00..1.01 rows=1 width=12)
         ->  Hash  (cost=1.08..1.08 rows=5 width=12)
               ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..1.08 rows=5 width=12)
                     ->  Seq Scan on foo  (cost=0.00..1.02 rows=2 width=12)
 Optimizer: Postgres-based planner
(14 rows)

drop index idx_bar_PT_ab;
drop index idx_bar_PT_ba;
-- 1.4 Test case for dynamic index only scans
CREATE INDEX idx_bar_PT_abc ON bar_PT USING btree(a,b,c);
CREATE INDEX idx_bar_PT_cba ON bar_PT USING btree(c,b,a);
vacuum analyze bar_PT;
    --  Index idx_bar_PT_abc should be selected
explain select * from bar_PT join foo on bar_PT.a =foo.a;
                                        QUERY PLAN                                         
-------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=1.04..7.23 rows=5 width=24)
   ->  Hash Join  (cost=1.04..7.16 rows=2 width=24)
         Hash Cond: (bar_pt_1_prt_1.a = foo.a)
         ->  Append  (cost=0.00..6.09 rows=6 width=12)
               Partition Selectors: $0
               ->  Seq Scan on bar_pt_1_prt_1  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_2  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_3  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_4  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_5  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_6  (cost=0.00..1.01 rows=1 width=12)
         ->  Hash  (cost=1.02..1.02 rows=2 width=12)
               ->  Partition Selector (selector id: $0)  (cost=0.00..1.02 rows=2 width=12)
                     ->  Seq Scan on foo  (cost=0.00..1.02 rows=2 width=12)
 Optimizer: Postgres-based planner
(15 rows)

    --  Index idx_bar_PT_cba should be selected
explain select * from bar_PT join foo on bar_PT.c =foo.c;
                                            QUERY PLAN                                            
--------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=1.15..7.34 rows=5 width=24)
   ->  Hash Join  (cost=1.15..7.27 rows=2 width=24)
         Hash Cond: (bar_pt_1_prt_1.c = foo.c)
         ->  Append  (cost=0.00..6.09 rows=6 width=12)
               ->  Seq Scan on bar_pt_1_prt_1  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_2  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_3  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_4  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_5  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_6  (cost=0.00..1.01 rows=1 width=12)
         ->  Hash  (cost=1.08..1.08 rows=5 width=12)
               ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..1.08 rows=5 width=12)
                     ->  Seq Scan on foo  (cost=0.00..1.02 rows=2 width=12)
 Optimizer: Postgres-based planner
(14 rows)

drop table if exists foo;
drop table if exists bar_PT;
------------------------------------------------
-- Scenario2: Unindexed predicate column in index.
------------------------------------------------
-- If the index does not cover all the columns of the predicate, than a scan cost
-- using it should be higher compared to an index which covers all the predicate columns.
drop table if exists foo;
NOTICE:  table "foo" does not exist, skipping
create table foo (a int , b int, c int) distributed by (a);
insert into foo select i,i,i from generate_series(1,5)i;
-- Adding 1,1,1 in the table so that we have different values for NDV and table rows.
insert into foo select 1,1,1*i/i from generate_series(1,5)i;
-- 2.1 Test case for index scans
CREATE INDEX idx_foo_a ON foo USING btree(a);
CREATE INDEX idx_foo_abc ON foo USING btree(a,b,c);
vacuum analyze foo;
-- Query1 - Index idx_foo_a should be selected.
explain select * from foo where a=1;
                                 QUERY PLAN                                  
-----------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..1.12 rows=6 width=12)
   ->  Seq Scan on foo  (cost=0.00..1.04 rows=2 width=12)
         Filter: (a = 1)
 Optimizer: Postgres-based planner
(4 rows)

-- Query2 - Index idx_foo_abc should be selected.
explain select * from foo where a=1 and b=1 and c=1;
                                 QUERY PLAN                                  
-----------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..1.11 rows=4 width=12)
   ->  Seq Scan on foo  (cost=0.00..1.06 rows=1 width=12)
         Filter: ((a = 1) AND (b = 1) AND (c = 1))
 Optimizer: Postgres-based planner
(4 rows)

drop index idx_foo_a;
drop index idx_foo_abc;
-- 2.2 Test case for dynamic index scans
drop table if exists foo;
drop table if exists bar_PT;
NOTICE:  table "bar_pt" does not exist, skipping
create table foo (a int , b int, c int) distributed by (a);
insert into foo select i,i,i from generate_series(1,5)i;
analyze foo;
    -- Partitioned Table
create table bar_PT (a int, b int, c int) partition by range(a) (start (1) inclusive end (12) every (2)) distributed by (a);
insert into bar_PT select i,i,i from generate_series(1,11)i;
CREATE INDEX idx_bar_PT_a ON bar_PT USING btree(a);
CREATE INDEX idx_bar_PT_abc ON bar_PT USING btree(a,b,c);
vacuum analyze bar_PT;
-- Query1: Index idx_bar_PT_a should be selected.
explain select * from bar_PT join foo on bar_PT.a =foo.a ;
                                        QUERY PLAN                                         
-------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=1.04..7.23 rows=5 width=24)
   ->  Hash Join  (cost=1.04..7.16 rows=2 width=24)
         Hash Cond: (bar_pt_1_prt_1.a = foo.a)
         ->  Append  (cost=0.00..6.09 rows=6 width=12)
               Partition Selectors: $0
               ->  Seq Scan on bar_pt_1_prt_1  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_2  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_3  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_4  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_5  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_6  (cost=0.00..1.01 rows=1 width=12)
         ->  Hash  (cost=1.02..1.02 rows=2 width=12)
               ->  Partition Selector (selector id: $0)  (cost=0.00..1.02 rows=2 width=12)
                     ->  Seq Scan on foo  (cost=0.00..1.02 rows=2 width=12)
 Optimizer: Postgres-based planner
(15 rows)

-- Query2 : Index idx_bar_PT_abc should be selected.
explain select * from bar_PT join foo on bar_PT.a =foo.a and bar_PT.b =foo.b and bar_PT.c =foo.c;
                                                  QUERY PLAN                                                   
---------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=1.05..7.25 rows=3 width=24)
   ->  Hash Join  (cost=1.05..7.20 rows=1 width=24)
         Hash Cond: ((bar_pt_1_prt_1.a = foo.a) AND (bar_pt_1_prt_1.b = foo.b) AND (bar_pt_1_prt_1.c = foo.c))
         ->  Append  (cost=0.00..6.09 rows=6 width=12)
               Partition Selectors: $0
               ->  Seq Scan on bar_pt_1_prt_1  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_2  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_3  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_4  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_5  (cost=0.00..1.01 rows=1 width=12)
               ->  Seq Scan on bar_pt_1_prt_6  (cost=0.00..1.01 rows=1 width=12)
         ->  Hash  (cost=1.02..1.02 rows=2 width=12)
               ->  Partition Selector (selector id: $0)  (cost=0.00..1.02 rows=2 width=12)
                     ->  Seq Scan on foo  (cost=0.00..1.02 rows=2 width=12)
 Optimizer: Postgres-based planner
(15 rows)

drop table if exists foo;
drop table if exists bar_PT;
